whylog/constraints/__init__.py

```diff
@@ -1,52 +1,202 @@
-from abc import ABCMeta
+from abc import ABCMeta, abstractmethod, abstractproperty
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstructorGroupsCountError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
-    pass
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    @abstractproperty
+    def TYPE(self):
+        pass
 
+    @abstractproperty
+    def MIN_GROUPS_COUNT(self):
+        return None
 
-class TimeConstraint(AbstractConstraint):
-    def __init__(self, groups, param_dict):
+    @abstractproperty
+    def MAX_GROUPS_COUNT(self):
+        return None
+
+    @abstractproperty
+    def PARAMS(self):
         pass
 
-    @classmethod
-    def verify(cls, group_contents, param_dict):
-        # TODO remove mock
-        return True
+    @abstractmethod
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
+        """
+        For Teacher and Front use while creating user rule.
 
+        :param param_dict: dict of additional params of constraint
+        :param groups: all groups that are linked by constraint,
+                       represented by list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
+        """
+        self.groups = groups or []
+        self.params = param_dict or {}
+        if params_checking:
+            self._check_constructor_groups()
+            self._check_constructor_params()
+
+    def _check_constructor_groups(self):
+        groups_count = len(self.groups)
+        if (self.MIN_GROUPS_COUNT is not None and groups_count < self.MIN_GROUPS_COUNT) \
+                or (self.MAX_GROUPS_COUNT is not None and groups_count > self.MAX_GROUPS_COUNT):
+            raise ConstructorGroupsCountError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
+            )
+
+    def _check_constructor_params(self):
+        correct_param_names = set(self.get_param_names())
+        actual_param_names = set(self.params.keys())
+        self._check_useless_params(correct_param_names, actual_param_names)
+        self._check_mandatory_params(correct_param_names, actual_param_names)
+        self._check_optional_params(correct_param_names, actual_param_names)
+
+    def _check_useless_params(self, correct_param_names, actual_param_names):
+        if actual_param_names - correct_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, actual_param_names)
+
+    def _check_mandatory_params(self, correct_param_names, actual_param_names):
+        pass
```

Comment:
this is a method in an abstract class and it is here for a purpose. But as an author of subclass, I don't know it. Please add a docstring.